src/charm.py

```diff
@@ -213,74 +146,69 @@ def get_manifest(self):
             }
         )
 
-        # Kubernetes won't automatically restart the pod when the configmap changes
-        # unless we manually add the hash somewhere into the Deployment spec, so that
-        # it changes whenever the configmap changes.
-        config_hash = sha256()
-        config_hash.update(config.encode("utf-8"))
-
-        context = {
-            "name": self.model.app.name.replace("-operator", ""),
-            "namespace": self.model.name,
-            "port": self.model.config["port"],
-            "config_yaml": config,
-            "config_hash": config_hash.hexdigest(),
-        }
+        # Get current layer
+        current_layer = self._container.get_plan()
+        # Create a new config layer
+        new_layer = self._dex_auth_layer
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self._container.add_layer(self._container_name, new_layer, combine=True)
+            self.logger.info("Pebble plan updated with new configuration")
 
-        return [
-            obj
-            for path in glob("src/manifests/*.yaml")
-            for obj in codecs.load_all_yaml(Path(path).read_text(), context=context)
-        ]
+        # Get current dex config
+        current_config = self._container.pull(self._dex_config_path).read()
+        if current_config != config:
```

Comment:
I get your point, though this is a very specific example. 
The first time dex starts the container, it pushes the default [config file](https://github.com/dexidp/dex/blob/v2.31.x/config.docker.yaml), which `yaml.safe_load` fails to parse:
```
    current_config = yaml.safe_load(self._container.pull(self._dex_config_path).read())
  File "/var/lib/juju/agents/unit-dex-auth-0/charm/venv/yaml/__init__.py", line 162, in safe_load
    return load(stream, SafeLoader)
    raise ParserError("while parsing a %s node" % node, start_mark,
  yaml.parser.ParserError: while parsing a flow node
  expected the node content, but found '-'
```
To handle it, the yaml would have to be 'sanitized' first, so it makes it less reliable.
While comparing two yamls, the config gets updated only once, when it overwrites the default file on install, and then only if config changed.
So I think it's ok to compare yamls instead of dicts in this case, please correct me if I'm wrong.

Reply:
Yeh I think that's fine.

Other option is: you know if the pebble plan is empty, and thus you've not pushed a config yet so on the first plan you could just blindly push the config, then subsequently compare dicts.

But either is fine, thanks for the explanation :)